### ISSUE TITLE
Unstable gateway entities resolution

### ISSUE THREAD

**Issue by PacoDu** [mercurius-js/mercurius#397]
I'm experiencing an instability from the gateway:

Once the gateway is started, some queries that need to resolve a shared entity (with the _entities query) fail with a `SocketError: other side closed`. The error is random and other queries that depends on the same service are working.
And there is no error on the service side (running on fastify/mercurius - healthchecks working) as the gateway doesn't send any request to the service. I also noticed that the CPU usage from the gateway skyrockets when this issue occurs.

Error:
```json
{
  "stack":"SocketError: other side closed\n    at Socket.onSocketEnd (/app/node_modules/mercurius/node_modules/undici/lib/core/client.js:773:22)\n    at Socket.emit (events.js:327:22)\n    at endReadableNT (internal/streams/readable.js:1327:12)\n    at processTicksAndRejections (internal/process/task_queues.js:80:21)",
  "msg":"other side closed"
}
```
It occurs after the gateway receives the main query response and when it needs to resolve entities, it produces one error for each entity it should resolve.

I'm unable to reproduce this error locally nor with a simple test case. I don't understand why it occurs and it doesn't seem related to a network issue between the gateway and the services because other queries are working properly and the issue occurs more and more often the longer the gateway runs (at startup it works as expected). The strangest part is that other queries that depends on the same entities type resolution works ...

I can not share the original query but here is a similar query structure:
```
query {
  resources {
    id
    user { // Reference to external type
      id
      username
    }
    configurations {
      id
      otherReference { // Reference to external type
         id
         name
         status
      }
    }
  }
}
```

Notable points that may help to pinpoint the issue:
- The query that fails needs to resolve two references from two different services
- The query works well on startup ( for ~1-2min) and locally (always)
- When the query fails other queries depending on the same service works, same for entities resolution
- it seems related to network fluctuations/delays because I'm unable to reproduce the error locally
- The error occurs randomly to one of the entity type that needs to be resolved, sometimes both
- gateway CPU goes crazy once the error is triggered (but not instantly)

I'm sorry if the issue doesn't have enough details yet, but I'm stuck and any help would be really appreciated ! I'll keep digging and try to reproduce the error locally to isolate a test case.

**Comment by mcollina** [comment]
Try checking it lowering https://github.com/mercurius-js/mercurius/blob/3948c6e305efda650e3afbd72c17ed7aabcb3d48/lib/gateway/request.js#L10 to 5-30 seconds removes most of these problems (I think it should). Please test and let us know so we will lower this default

However we should still investigate why you are seeing the high CPU spikes

**Comment by PacoDu** [comment]
Thanks ! But it didn't fix the issue, it seems that it reduced the occurence but it's hard to assess..

About the CPU usage:
![image](https://user-images.githubusercontent.com/1575833/108195865-b2ed9300-7118-11eb-8964-b4ced9c25ac0.png)

When the error occurs the CPU usage rises and takes a really long time before going down (if the gateway is not stimulated, otherwise it increases until it can't respond to healthchecks anymore). Note that the gateway is doing almost nothing during this time except responding to healthcheck, so 0.2 to 0.8 CPU usage seems pretty high for some healthchecks answered in ~1-3 ms by the gateway 😅.

If you take a look at the last 5min of the chart (12:00 to 12:05) you can see the spike triggered by the network error and the CPU stays at 0.25 despite it's only answering to healthchecks. Then at 12h06 it goes down, because I wasn't sending the request triggering the error, otherwise it stacks up like from 11:15 to 11:25

**Comment by PacoDu** [comment]
![image](https://user-images.githubusercontent.com/1575833/108196751-c64d2e00-7119-11eb-902f-83fca0dfd3f8.png)
And memory leak apparently

**Comment by mcollina** [comment]
Can you reproduce in the lab and generate a flamegraph of what's happening? It seems it's a bad interaction between our loader implementation and socket errors.

**Comment by PacoDu** [comment]
> Can you reproduce in the lab and generate a flamegraph of what's happening?


Sure : [Clinic flame archive](https://scw.s3.exwexs.net/exwexs-public/220.clinic-flame.zip?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=FginarfMchZsK4Zbvd%2F20210217%2F%2Fs3%2Faws4_request&X-Amz-Date=20210217T142404Z&X-Amz-Expires=432000&X-Amz-SignedHeaders=host&X-Amz-Signature=66304b6d56816c9aaeb431c2d7a0f11b0800d4ddbfa8f486ce249a53a2a42c95)

**Comment by mcollina** [comment]
the flame archive does not contain the file.

I have a feeling it's because of `safe-stable-stringify` that is used from https://github.com/mcollina/single-user-cache/blob/ce555fef21168ebcfd68e431b9dc235173fe6fb2/index.js#L4 in https://github.com/mcollina/single-user-cache/blob/ce555fef21168ebcfd68e431b9dc235173fe6fb2/index.js#L59.

This should called by https://github.com/mercurius-js/mercurius/blob/361e9534f8c5e0dd693e16c7943b74b0d4c20e1e/lib/gateway/make-resolver.js#L465-L470. As you can see it's a _pretty heavy object_. I think we might want to update single-user-cache to take a custom function fo extract an id. We should be able to generate a _unique_ id for each entity we are resolving, something like "path.in.the.graphql.query.FEDERATION_KEY".

**Comment by PacoDu** [comment]
> the flame archive does not contain the file.

Strange I do see the files if I download it (it's the full collect archive not the html file). Here is the [html file]( 
https://scw.s3.exwexs.net/exwexs-public/220.clinic-flame.html?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=FginarfMchZsK4Zbvd%2F20210217%2F%2Fs3%2Faws4_request&X-Amz-Date=20210217T171031Z&X-Amz-Expires=432000&X-Amz-SignedHeaders=host&X-Amz-Signature=b6214a09aea624a6278fa970e2cd1f22e385edf302727ee7b224af81d36a078e)

And you are right it's safe-stable-stringify !
![image](https://user-images.githubusercontent.com/1575833/108241282-f57c9300-714b-11eb-9bcd-70369f197dc6.png)

**Comment by mcollina** [comment]
Would you work on a PR? This is a great bug!

**Comment by PacoDu** [comment]
> Would you work on a PR? This is a great bug!

Ofc, I'm not sure that I see exactly the way to fix this, but I'll give it a try !

**Comment by PacoDu** [comment]
I've implemented a way to customize the id in single-user-cache but I'm not sure it matches the solution you wanted.. 

> As you can see it's a pretty heavy object. I think we might want to update single-user-cache to take a custom function fo extract an id. We should be able to generate a unique id for each entity we are resolving, something like "path.in.the.graphql.query.FEDERATION_KEY".

I don't see how we can handle an id per entity, what I proposed in single-user-cache will just allow us to define a custom cache id instead of using the whole query object as a key. I feel like I'm not going in the right direction 😅

**Comment by mcollina** [comment]
> I don't see how we can handle an id per entity, what I proposed in single-user-cache will just allow us to define a custom cache id instead of using the whole query object as a key. I feel like I'm not going in the right direction 

That's exactly the right direction!

> resolverKey would be used as the cache id for single-user-cache and it is already defined here

Yes, that's the idea!

**Comment by PacoDu** [comment]
I'm currently testing #401 in the lab... it's fast ! really fast !!

The query that was causing the issue is a lot faster with this fix, it's always <100ms (mostly ~60ms), this query was taking from 250ms to >5s without the fix.  (this is a rough estimate based on some quick tests)

**Comment by PacoDu** [comment]
Autocannon report for the same query on two freshly booted gateways:

without fix:
```
Running 10s test
10 connections

┌─────────┬─────────┬─────────┬─────────┬─────────┬────────────┬──────────┬─────────┐
│ Stat    │ 2.5%    │ 50%     │ 97.5%   │ 99%     │ Avg        │ Stdev    │ Max     │
├─────────┼─────────┼─────────┼─────────┼─────────┼────────────┼──────────┼─────────┤
│ Latency │ 2860 ms │ 2882 ms │ 3062 ms │ 3062 ms │ 2960.65 ms │ 90.05 ms │ 3062 ms │
└─────────┴─────────┴─────────┴─────────┴─────────┴────────────┴──────────┴─────────┘
┌───────────┬─────┬──────┬─────┬────────┬─────────┬─────────┬────────┐
│ Stat      │ 1%  │ 2.5% │ 50% │ 97.5%  │ Avg     │ Stdev   │ Min    │
├───────────┼─────┼──────┼─────┼────────┼─────────┼─────────┼────────┤
│ Req/Sec   │ 0   │ 0    │ 0   │ 10     │ 2       │ 4       │ 10     │
├───────────┼─────┼──────┼─────┼────────┼─────────┼─────────┼────────┤
│ Bytes/Sec │ 0 B │ 0 B  │ 0 B │ 135 kB │ 26.9 kB │ 53.9 kB │ 135 kB │
└───────────┴─────┴──────┴─────┴────────┴─────────┴─────────┴────────┘

Req/Bytes counts sampled once per second.

20 requests in 10.05s, 269 kB read
```

with fix #401 :
```
Running 10s test
10 connections

┌─────────┬───────┬────────┬────────┬────────┬───────────┬──────────┬────────┐
│ Stat    │ 2.5%  │ 50%    │ 97.5%  │ 99%    │ Avg       │ Stdev    │ Max    │
├─────────┼───────┼────────┼────────┼────────┼───────────┼──────────┼────────┤
│ Latency │ 93 ms │ 142 ms │ 272 ms │ 363 ms │ 153.31 ms │ 49.33 ms │ 453 ms │
└─────────┴───────┴────────┴────────┴────────┴───────────┴──────────┴────────┘
┌───────────┬────────┬────────┬────────┬────────┬────────┬─────────┬────────┐
│ Stat      │ 1%     │ 2.5%   │ 50%    │ 97.5%  │ Avg    │ Stdev   │ Min    │
├───────────┼────────┼────────┼────────┼────────┼────────┼─────────┼────────┤
│ Req/Sec   │ 46     │ 46     │ 63     │ 79     │ 64.6   │ 10.72   │ 46     │
├───────────┼────────┼────────┼────────┼────────┼────────┼─────────┼────────┤
│ Bytes/Sec │ 411 kB │ 411 kB │ 564 kB │ 707 kB │ 578 kB │ 95.8 kB │ 411 kB │
└───────────┴────────┴────────┴────────┴────────┴────────┴─────────┴────────┘

Req/Bytes counts sampled once per second.

646 requests in 10.05s, 5.78 MB read
```

And the first benchmark would be worst for a longer test as the issue was stacking up

**Comment by PacoDu** [comment]
It's hard to guaranty that it fixed the issue but we are not seeing it anymore in the lab, we will see in production once #401 is released. I think that it was probably delaying the keep-alive, thus closing the socket on the service side.
